File: AlgorithmStudy/10주차/BOJ1406/김재환1406.py
import sys
input = sys.stdin.readline

# 초기 문자열
initial = input().rstrip()

# 왼쪽/오른쪽 스택
left = list(initial)  # 커서 왼쪽
right = []            # 커서 오른쪽

# 명령 수
M = int(input().rstrip())

for _ in range(M):
    cmd = input().split()

    # 커서 왼쪽
    if cmd[0] == 'L':
        if left:  # 왼쪽이 비어 있지 않으면
            right.append(left.pop())

    # 커서 오른쪽
    elif cmd[0] == 'D':
        if right:  # 오른쪽이 비어 있지 않으면
            left.append(right.pop())

    # 왼쪽 문자 삭제
    elif cmd[0] == 'B':
        if left:
            left.pop()

    # 삽입
    elif cmd[0] == 'P':
        # cmd[1]에 넣을 문자
        left.append(cmd[1])

# 출력: left + reversed(right)
# reversed(right)는 스택이니까 뒤집어야 원래 순서
print(''.join(left + right[::-1]))

File: AlgorithmStudy/10주차/BOJ1406/김재환1406.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;
import java.util.ArrayDeque;
import java.util.Deque;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        // 초기 문자열
        String initial = br.readLine();

        // 왼쪽/오른쪽 덱
        Deque<Character> left = new ArrayDeque<>();
        Deque<Character> right = new ArrayDeque<>();

        // 초기 문자열의 문자들을 left에 삽입
        for (char c : initial.toCharArray()) {
            left.addLast(c);
        }

        int M = Integer.parseInt(br.readLine());
        while (M-- > 0) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            String cmd = st.nextToken();

            if (cmd.equals("L")) {
                if (!left.isEmpty()) {
                    right.addLast(left.removeLast());
                }
            } else if (cmd.equals("D")) {
                if (!right.isEmpty()) {
                    left.addLast(right.removeLast());
                }
            } else if (cmd.equals("B")) {
                if (!left.isEmpty()) {
                    left.removeLast();
                }
            } else if (cmd.equals("P")) {
                char x = st.nextToken().charAt(0);
                left.addLast(x);
            }
        }

        // left에는 앞쪽 순서, right에는 뒤집힌 커서 오른쪽
        StringBuilder sb = new StringBuilder();

        for (char c : left) {
            sb.append(c);
        }

        // right는 뒤집어서 붙임
        while (!right.isEmpty()) {
            sb.append(right.removeLast());
        }

        System.out.print(sb.toString());
    }
}
